pkg/library/home/persistentHome.go

```diff
@@ -95,3 +99,77 @@ func storageStrategySupportsPersistentHome(workspace *common.DevWorkspaceWithCon
 	storageClass := workspace.Spec.Template.Attributes.GetString(constants.DevWorkspaceStorageTypeAttribute, nil)
 	return storageClass != constants.EphemeralStorageClassType
 }
+
+func addInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec, initContainerFromConfig *v1alpha2.Container) {
+
+	initContainerComponent := v1alpha2.Component{
+		Name: constants.HomeInitComponentName,
+	}
+
+	initContainer := initContainerFromConfig
+
+	if initContainer == nil {
+		initContainer = inferInitContainer(dwTemplateSpec)
+	}
+
+	if initContainer == nil {
+		return
+	}
+
+	initContainerComponent.ComponentUnion = v1alpha2.ComponentUnion{
+		Container: &v1alpha2.ContainerComponent{
+			Container: *initContainer,
+		},
+	}
+
+	dwTemplateSpec.Components = append(dwTemplateSpec.Components, initContainerComponent)
+	dwTemplateSpec.Commands = append(dwTemplateSpec.Commands, v1alpha2.Command{
+		Id: constants.HomeInitEventId,
+		CommandUnion: v1alpha2.CommandUnion{
+			Apply: &v1alpha2.ApplyCommand{
+				Component: constants.HomeInitComponentName,
+			},
+		},
+	})
+	dwTemplateSpec.Events.PreStart = append(dwTemplateSpec.Events.PreStart, constants.HomeInitEventId)
+}
+
+func inferInitContainer(dwTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec) *v1alpha2.Container {
+	var nonImportedComponent v1alpha2.Component
+	for _, component := range dwTemplateSpec.Components {
+		if component.Container == nil {
+			continue
+		}
+
+		pluginSource := component.Attributes.GetString(constants.PluginSourceAttribute, nil)
+		if pluginSource == "" || pluginSource == "parent" {
+			// First, non-imported container component is selected
+			nonImportedComponent = component
+			break
+		}
+	}
+
+	createOutFile := "touch /home/user/.init-persistent-home"
+	runSetupScript := `([ -n "$HOME_SETUP_SCRIPT" ] && [ -f "$HOME_SETUP_SCRIPT" ] && echo "Running $HOME_SETUP_SCRIPT" && (source "$HOME_SETUP_SCRIPT" || true))`
+	runStow := `
+	(command -v stow > /dev/null && \
+		STOW_COMPLETE=/home/user/.stow_completed && \
+		[ ! -f $STOW_COMPLETE ] && \
+		echo "Running/ stow" && \
+		stow . -t /home/user/ -d /home/tooling/ --no-folding --adopt -v 2 > /tmp/stow.log 2>&1 && \
+		cp /home/tooling/.viminfo /home/user/.viminfo && \
```

Comment:
It's probably safer to use `cp -n` for `cp --no-clobber`, which will cancel the cp operator if the destination file already exists. It's rare that this will ever be needed, since the home persistence setup is only supposed to happen the first time a workspace starts, but it shouldn't hurt.

I should have used `cp -n` in the UDI's entrypoint from the start, to be honest.

Reply:
Thanks, fixed